### PR TITLE
CMP-2132: Implement suspend and resume scan schedule

### PR DESCRIPTION
**Pull request by rhmdnd**
This commit implements the logic and tests necessary to suspend and
resume scan schedules using the `ScanSetting` custom resource.

You can find more details on the overall justification, use cases, and
implementation details in the enhancement:

  https://github.com/ComplianceAsCode/compliance-operator/pull/375

### CODE REVIEW

File: tests/e2e/parallel/main_test.go [hunk 1]
```diff
@@ -2795,3 +2797,174 @@ func TestScheduledSuiteTimeoutFail(t *testing.T) {
...
+			Kind:     "ScanSetting",
+			APIGroup: "compliance.openshift.io/v1alpha1",
+		},
+	}
+	if err := f.Client.Create(context.TODO(), &scanSettingBinding, nil); err != nil {
+		t.Fatal(err)
+	}
+	defer f.Client.Delete(context.TODO(), &scanSettingBinding)
+
+	// Assert that no scan takes place
```
Comment (rhmdnd) [inline tests/e2e/parallel/main_test.go:2959]: This assertion and the corresponding implementation in the compliance suite controller needs to be implemented.

File: CHANGELOG.md [hunk 2]
```diff
@@ -9,7 +9,12 @@ Versioning](https://semver.org/spec/v2.0.0.html).
 
 ### Enhancements
 
--
+- Users can now pause scan schedules by setting the `ScanSetting.suspend`
```
Comment (sheriff-rh) [inline CHANGELOG.md:12]: This seems like a great QOL feature to add! I'll be sure it's highlighted in the release notes downstream in the future.

**Review by sheriff-rh — COMMENTED** [review]
LGTM!

File: tests/e2e/parallel/main_test.go [hunk 3]
```diff
@@ -2795,3 +2798,183 @@ func TestScheduledSuiteTimeoutFail(t *testing.T) {
...
+	if *job.Spec.Suspend == true {
+		t.Fatalf("CronJob %s is suspended when it should not be", jobName)
+	}
+
+	// Suspend the `ScanSetting` using the `suspend` attribute
+	scanSettingUpdate := &compv1alpha1.ScanSetting{}
+	if err := f.Client.Get(context.TODO(), types.NamespacedName{Namespace: f.OperatorNamespace, Name: scanSettingName}, scanSettingUpdate); err != nil {
+		t.Fatalf("failed to get ScanSetting %s", scanSettingName)
+	}
+	scanSettingUpdate.Suspend = true
```
Comment (rhmdnd) [inline tests/e2e/parallel/main_test.go:2883]: For some reason in my local testing, this isn't actually updating the `CronJob`.

File: pkg/controller/compliancesuite/suitererunner_cron_compat.go
Comment (Vincent056) [inline pkg/controller/compliancesuite/suitererunner_cron_compat.go:1]: I think you also need to edit line 65

Comment (Vincent056) [reply pkg/controller/compliancesuite/suitererunner_cron_compat.go:1]: I think you might need to skip the update if they are the same:

Change:
```
		if getObjTyped.Spec.Schedule == suite.Spec.Schedule {
			return nil
		}
```
to
```
		if getObjTyped.Spec.Schedule == suite.Spec.Schedule &&  getObjTyped.Spec.Suspend == suite.Spec.Suspend{
			return nil
		}
```

Comment (rhmdnd) [reply pkg/controller/compliancesuite/suitererunner_cron_compat.go:1]: Aha - nice catch!

File: tests/e2e/parallel/main_test.go [hunk 4]
```diff
@@ -2795,3 +2797,189 @@ func TestScheduledSuiteTimeoutFail(t *testing.T) {
...
+	if err != nil {
+		t.Fatal(err)
+	}
+	if *job.Spec.Suspend == true {
+		t.Fatalf("CronJob %s is suspended when it should not be", jobName)
+	}
+
+}
+
+func TestSuspendScanSettingDoesNotCreateScan(t *testing.T) {
```
Comment (rhmdnd) [inline tests/e2e/parallel/main_test.go:2921]: Maybe we should move these to serial tests since they will fail if they're run at the same time as another test that's using `ocp4-cis` or `ocp4-cis-node`.

**Comment by rhmdnd** [comment]
@xiaojiey I have one minor test fix to make, but otherwise these should be good to go for pre-verification.

File: tests/e2e/parallel/main_test.go [hunk 5]
```diff
@@ -2795,3 +2797,189 @@ func TestScheduledSuiteTimeoutFail(t *testing.T) {
...
+	// Assert the `CronJob` associated with the `ComplianceSuite` is set to `suspend=true`
+	job = &batchv1.CronJob{}
+	jobName = compsuitectrl.GetRerunnerName(suite.Name)
+	err = f.Client.Get(context.TODO(), types.NamespacedName{Name: jobName, Namespace: f.OperatorNamespace}, job)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if *job.Spec.Suspend == false {
+		t.Fatalf("Expected CronJob %s to be suspended", jobName)
+	}
```
Comment (yuumasato) [inline tests/e2e/parallel/main_test.go:2896]: If we add `SUSPENDED` Status/Phase, then here we would also check for the status to be `SUSPENDED` with`WaitForSuiteScansStatus()`.

File: tests/e2e/parallel/main_test.go [hunk 6]
```diff
@@ -2795,3 +2797,189 @@ func TestScheduledSuiteTimeoutFail(t *testing.T) {
...
+		t.Fatal(err)
+	}
+	defer f.Client.Delete(context.TODO(), &scanSettingBinding)
+
+	// Assert that no scan takes place
+	scanName := "ocp4-cis"
+	err := f.AssertScanDoesNotExist(scanName, f.OperatorNamespace)
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (yuumasato) [inline tests/e2e/parallel/main_test.go:2974]: Here as well, check the status of the ComplianceSuite.

**Comment by xiaojiey** [comment]
@rhmdnd  Generally it is quite good. 
Just two minor questions, 
1. Is it needed to set default value for suspend for the default and default-auto-apply ss? Currently the value is unset.
$ oc get ss default -o=jsonpath={.suspend}
$ oc get ss default-auto-apply -o=jsonpath={.suspend}
2. When creating a ss with suspend set to true, then create a ssb with the ss, the suite with be in PENDING status. I didn't notice if there is any warning message to know why it is PENDING. Is it possible to add such warning message? Thanks.
$ oc get suite
NAME   PHASE     RESULT
test   PENDING
```
All below scenarios verified with 4.14.0-0.nightly-2023-09-11-201102 + code in this PR 
Test scenario 1##########Create a ss with suspend set to false, then suspend the ssb on demand:
1. Create a ss with suspend to false, and target to scan every 3 minutes 
schedule: "*/3 * * * *"
suspend: false
2. Create a ssb:
$ oc compliance bind -N test -S test-suspend profile/ocp4-cis
Creating ScanSettingBinding test
$ oc get cronjob
NAME            SCHEDULE      SUSPEND   ACTIVE   LAST SCHEDULE   AGE
test-rerunner   */3 * * * *   False     0        108s            2m51s
$ oc get job
NAME                     COMPLETIONS   DURATION   AGE
test-rerunner-28241790   1/1           5s         114s
test-rerunner-28241793   0/1                      0s
$ oc get suite
NAME   PHASE     RESULT
test   RUNNING   NOT-AVAILABLE
4. Path the suspend to true:
$ oc patch ss test-suspend -p '{"suspend":true}' --type='merge'
scansetting.compliance.openshift.io/test-suspend patched
5. Check the cronjob will be supspended:
$ oc get job
NAME                     COMPLETIONS   DURATION   AGE
test-rerunner-28241790   1/1           5s         10m
test-rerunner-28241793   1/1           6s         7m5s
6. Patch to set the suspend to false:
$ oc patch ss test-suspend -p '{"suspend":false}' --type='merge'
scansetting.compliance.openshift.io/test-suspend patched
7. Check the cronjob:
$ oc get cronjob -w
NAME            SCHEDULE      SUSPEND   ACTIVE   LAST SCHEDULE   AGE
test-rerunner   */3 * * * *   False     0        3m              13m
test-rerunner   */3 * * * *   False     1        0s              13m
^C$ oc get suite -w
NAME   PHASE     RESULT
test   RUNNING   NOT-AVAILABLE
^C
$ oc get job
NAME                     COMPLETIONS   DURATION   AGE
test-rerunner-28241793   1/1           6s         9m31s
test-rerunner-28241799   1/1           7s         2m12s
test-rerunner-28241802   1/1           7s         31s

Test scenario 2##########Create a ss with suspend set to true, then create a ssb and unsuspend the ssb on demand:
1. Create a ss with suspend set to true
schedule: "*/3 * * * *"
suspend: true
2. Create a ssb:
$ oc apply -f ss2.yaml 
scansetting.compliance.openshift.io/test-suspend-true created
3. Check the suite will be in PENDING status:
$ oc get ssb
NAME   AGE
test   3s
$ oc get suite
NAME   PHASE     RESULT
test   PENDING
4. Patch the ss to set suspend to false
$ oc patch ss test-suspend-true -p '{"suspend":false}' --type='merge'
scansetting.compliance.openshift.io/test-suspend-true patched
5. Check the suite will be in running status immediately:
$ oc get suite
NAME   PHASE   RESULT
test   DONE    NON-COMPLIANT
$ oc get cronjob
NAME            SCHEDULE      SUSPEND   ACTIVE   LAST SCHEDULE   AGE
test-rerunner   */3 * * * *   False     0        114s            3m45s
$ oc get job
NAME                     COMPLETIONS   DURATION   AGE
test-rerunner-28241838   1/1           6s         118s


Test scenario 3##########Create a ss with suspend not set, then create a ssb suspend the ssb on demand:
1. create a ss:
$ oc get ss test-default -o=jsonpath={.suspend}
$ oc get ss test-default -o=jsonpath={.schedule}
*/3 * * * *
2. Create a ssb:
$ oc compliance bind -N test -S test-default profile/ocp4-cis
Creating ScanSettingBinding test
$ oc get suite -w
NAME   PHASE     RESULT
test   RUNNING   NOT-AVAILABLE
test   AGGREGATING   NOT-AVAILABLE
test   DONE          NON-COMPLIANT
test   DONE          NON-COMPLIANT
^C
$ oc get cronjob
NAME            SCHEDULE      SUSPEND   ACTIVE   LAST SCHEDULE   AGE
test-rerunner   */3 * * * *   False     0        2m9s            4m15s
3. Patch the suspend to true:
$ oc patch ss test-default -p '{"suspend":true}' --type='merge'
scansetting.compliance.openshift.io/test-default patched

$ oc get cronjob
NAME            SCHEDULE      SUSPEND   ACTIVE   LAST SCHEDULE   AGE
test-rerunner   */3 * * * *   True      0        8m54s           11m
$ oc get job
NAME                     COMPLETIONS   DURATION   AGE
test-rerunner-28241859   1/1           6s         8m57s
4. Check the cronjobs:
$ oc get cronjob 
NAME            SCHEDULE      SUSPEND   ACTIVE   LAST SCHEDULE   AGE
test-rerunner   */3 * * * *   False     0        67s             12m
$ oc get suite -w
NAME   PHASE     RESULT
test   RUNNING   NOT-AVAILABLE
^C$ oc get job
NAME                     COMPLETIONS   DURATION   AGE
test-rerunner-28241859   1/1           6s         10m
test-rerunner-28241868   1/1           6s         36s
```

File: pkg/apis/compliance/v1alpha1/compliancescan_types.go [hunk 7]
```diff
@@ -58,6 +58,8 @@ const (
 	PhaseRunning ComplianceScanStatusPhase = "RUNNING"
 	// PhaseAggregating represents the scan aggregating the results
 	PhaseAggregating ComplianceScanStatusPhase = "AGGREGATING"
+	// PhaseSuspended represents the scan schedule is on hold and will not run
+	PhaseSuspended ComplianceScanStatusPhase = "SUSPENDED"
```
Comment (rhmdnd) [inline pkg/apis/compliance/v1alpha1/compliancescan_types.go:62]: Revert this change since it makes more sense to associate the status to the binding instead of the suite. It makes it easier to short-circuit the suite creation if we keep associate the status to the binding.

File: pkg/controller/compliancesuite/compliancesuite_controller.go [hunk 8]
```diff
@@ -147,6 +147,13 @@ func (r *ReconcileComplianceSuite) Reconcile(ctx context.Context, request reconc
 		return reconcile.Result{}, r.issueValidationError(suite, errorMsg, reqLogger)
 	}
 
+	if r.isComplianceSuiteSuspended(suite) {
```
Comment (rhmdnd) [inline pkg/controller/compliancesuite/compliancesuite_controller.go:150]: Revert this since we're going to do this in the scansettingbinding controller instead.

File: pkg/controller/compliancesuite/compliancesuite_controller.go [hunk 9]
```diff
@@ -157,7 +164,10 @@ func (r *ReconcileComplianceSuite) Reconcile(ctx context.Context, request reconc
 		return reconcile.Result{}, nil
 	}
 
-	suiteCopy := suite.DeepCopy()
+	suiteCopy, err := r.refreshSuite(suite)
```
Comment (rhmdnd) [inline pkg/controller/compliancesuite/compliancesuite_controller.go:167]: We don't need this complexity if we have the status on the binding.

File: pkg/controller/compliancesuite/compliancesuite_controller.go [hunk 10]
```diff
@@ -171,18 +181,30 @@ func (r *ReconcileComplianceSuite) Reconcile(ctx context.Context, request reconc
 	}
 
 	if suiteCopy.IsResultAvailable() {
-		sCopy := suite.DeepCopy()
-		sCopy.Status.SetConditionReady()
-		updateErr := r.Client.Status().Update(context.TODO(), sCopy)
+		c := compv1alpha1.ComplianceSuite{}
```
Comment (rhmdnd) [inline pkg/controller/compliancesuite/compliancesuite_controller.go:184]: Similar comment as above.

File: pkg/controller/compliancesuite/compliancesuite_controller.go [hunk 11]
```diff
@@ -229,6 +229,13 @@ func (r *ReconcileComplianceSuite) issueValidationError(suite *compv1alpha1.Comp
 }
 
 func (r *ReconcileComplianceSuite) reconcileScans(suite *compv1alpha1.ComplianceSuite, logger logr.Logger) (bool, error) {
+	// Short-circuit and return without requeuing the request if the
+	// ComplianceSuite is suspended by the ScanSetting. When the
+	// ScanSetting.suspend attribute is disabled, or set to False, the
+	// ComplianceSuite will get updated and schedule scans.
+	if suite.Spec.Suspend {
+		return false, nil
```
[the remark this answers is not in the document]
Comment (rhmdnd) [reply pkg/controller/compliancesuite/compliancesuite_controller.go:287]: I got down the path of incorporating status into the ComplianceSuite, and realized that having a Suspended status there makes reconciling it when the `ScanSetting` is reactivated much more complicated.

I'll propose an alternative with the status on the `ScanSettingBinding`, which should allow us to relay the same information, but it makes it simpler to reconcile because we can short-circuit the `ComplianceScan` creation if we have a binding that references a suspended scan setting.

File: tests/e2e/framework/common.go [hunk 12]
```diff
@@ -2211,3 +2221,68 @@ func (f *Framework) WaitForGenericRemediationToBeAutoApplied(remName, remNamespa
 	}
 	return nil
 }
+
+func (f *Framework) AssertComplianceSuiteIsSuspended(name string) error {
```
Comment (rhmdnd) [inline tests/e2e/framework/common.go:2225]: We should be ok to remove this, and replace it with a similar function that asserts a binding is suspended.

File: tests/e2e/framework/common.go [hunk 13]
```diff
@@ -2211,3 +2221,68 @@ func (f *Framework) WaitForGenericRemediationToBeAutoApplied(remName, remNamespa
...
+		return err
+	}
+	if suite.Status.Phase != "DONE" {
+		msg := fmt.Sprintf("ComplianceSuite %s has phase %s, instead of \"DONE\"", suite.Name, string(suite.Status.Phase))
+		return errors.New(msg)
+	}
+	return nil
+}
+
+func (f *Framework) AssertCronJobIsSuspended(name string) error {
```
Comment (rhmdnd) [inline tests/e2e/framework/common.go:2264]: This will still be useful, regardless of where we put the suspended status.

File: pkg/controller/scansettingbinding/scansettingbinding_controller_test.go [hunk 14]
```diff
@@ -149,6 +149,7 @@ var _ = Describe("Testing scansettingbinding controller", func() {
 			ComplianceSuiteSettings: compv1alpha1.ComplianceSuiteSettings{
 				AutoApplyRemediations: true,
 				Schedule:              "0 1 * * *",
+				Suspend:               false,
```
Comment (rhmdnd) [inline pkg/controller/scansettingbinding/scansettingbinding_controller_test.go:152]: This will default to False now, so this should be removed.

File: tests/e2e/serial/main_test.go [hunk 15]
```diff
@@ -1408,6 +1409,183 @@ func TestKubeletConfigRemediation(t *testing.T) {
...
+	scanSettingName := framework.GetObjNameFromTest(t) + "-scansetting"
+	scanSetting := compv1alpha1.ScanSetting{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      scanSettingName,
+			Namespace: f.OperatorNamespace,
+		},
+		ComplianceSuiteSettings: compv1alpha1.ComplianceSuiteSettings{
+			AutoApplyRemediations: false,
+			Schedule:              "0 1 * * *",
+			Suspend:               false,
```
Comment (rhmdnd) [inline tests/e2e/serial/main_test.go:1425]: This isn't necessary since Suspend will default to `false`, but keeping it here helps readability.

File: pkg/apis/compliance/v1alpha1/scansettingbinding_types.go [hunk 16]
```diff
@@ -5,6 +5,15 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+type ScanSettingBindingStatusPhase string
```
Comment (rhmdnd) [inline pkg/apis/compliance/v1alpha1/scansettingbinding_types.go:8]: This could technically be pulled out into it's own patch.

File: pkg/controller/scansettingbinding/scansettingbinding_controller.go [hunk 17]
```diff
@@ -266,8 +315,11 @@ func (r *ReconcileScanSettingBinding) Reconcile(ctx context.Context, request rec
 	}
 
 	if scanSettingBindingStatusNeedsUpdate(instance) {
+		found := compliancev1alpha1.ComplianceSuite{}
+		err = r.Client.Get(context.TODO(), types.NamespacedName{Namespace: suite.Namespace, Name: suite.Name}, &found)
```
Comment (rhmdnd) [inline pkg/controller/scansettingbinding/scansettingbinding_controller.go:319]: Should do some error handling here.

I had to refetch the ComplianceSuite, otherwise we were updating a stale reference.

File: pkg/controller/scansettingbinding/scansettingbinding_controller.go [hunk 18]
```diff
@@ -756,7 +808,28 @@ func isGvk(obj *unstructured.Unstructured, expectGvk *schema.GroupVersionKind) e
...
+		}
+		log.Info("Found ScanSetting", "ScanSetting", scanSetting.Name)
+		if scanSetting.ComplianceSuiteSettings.Suspend != found.Spec.ComplianceSuiteSettings.Suspend {
+			log.Info("Detected change between ScanSetting and ComplianceSuite", "ScanSetting", scanSetting, "ComplianceSuite", found.Name)
+			return true
+		} else if scanSetting.ComplianceSuiteSettings.Schedule != found.Spec.ComplianceSuiteSettings.Schedule {
+			log.Info("Detected change between ScanSetting and ComplianceSuite", "ScanSetting", scanSetting, "ComplianceSuite", found.Name)
+			return true
+		}
+	}
```
Comment (Vincent056) [inline pkg/controller/scansettingbinding/scansettingbinding_controller.go:831]: I wonder if we need this because, we have 
https://github.com/ComplianceAsCode/compliance-operator/blob/cd3744f61a483d2d04fd79c3eac28e0d026005de/pkg/controller/scansettingbinding/scansettingbinding_controller.go#L339,
the suite should already have expected scansetthing in the have.Spec.ComplianceSuiteSettings

File: pkg/controller/scansettingbinding/scansettingbinding_controller.go [hunk 19]
```diff
@@ -246,7 +281,21 @@ func (r *ReconcileScanSettingBinding) Reconcile(ctx context.Context, request rec
 	}
 
 	// The suite already exists, should we update?
-	if suiteNeedsUpdate(&suite, &found) {
+	if r.suiteNeedsUpdate(&suite, &found, instance) {
```
Comment (Vincent056) [inline pkg/controller/scansettingbinding/scansettingbinding_controller.go:284]: Same as below, I feel we shouldn't need these because suite should already has latest scansetting from  https://github.com/ComplianceAsCode/compliance-operator/pull/396/files#diff-c647a52d3b0966655cdd2384a18bfdaa5e8d694df700395aabba380140b89ef2R211

File: pkg/controller/scansettingbinding/scansettingbinding_controller.go [hunk 20]
```diff
@@ -246,7 +281,21 @@ func (r *ReconcileScanSettingBinding) Reconcile(ctx context.Context, request rec
...
+		if instance.SettingsRef != nil {
+			scanSettings := &compliancev1alpha1.ScanSetting{}
+			err := r.Client.Get(context.TODO(), types.NamespacedName{Name: instance.SettingsRef.Name, Namespace: instance.Namespace}, scanSettings)
+			if err != nil {
+				// What should happen here?
+			}
+			suite.Spec.ComplianceSuiteSettings.Suspend = scanSettings.ComplianceSuiteSettings.Suspend
+			suite.Spec.ComplianceSuiteSettings.Schedule = scanSettings.ComplianceSuiteSettings.Schedule
+		}
+
```
Comment (Vincent056) [inline pkg/controller/scansettingbinding/scansettingbinding_controller.go:297]: same as below, I think we should already have the latest scansetting in the suite because https://github.com/ComplianceAsCode/compliance-operator/pull/396/files#diff-c647a52d3b0966655cdd2384a18bfdaa5e8d694df700395aabba380140b89ef2R211

File: tests/e2e/framework/common.go [hunk 21]
```diff
@@ -1014,6 +1047,28 @@ func (f *Framework) AssertScanHasValidPVCReferenceWithSize(scanName, size, names
...
+func (f *Framework) AssertComplianceSuiteDoesNotExist(name, namespace string) error {
+	cs := &compv1alpha1.ComplianceSuite{}
+	defer f.logContainerOutput(namespace, name)
+	err := f.Client.Get(context.TODO(), types.NamespacedName{Name: name, Namespace: namespace}, cs)
+	if apierrors.IsNotFound(err) {
+		return nil
+	} else if err != nil {
+		return err
+	}
+	return errors.New("Failed to assert ComplianceSuite %s does not exist.")
```
Comment (rhmdnd) [inline tests/e2e/framework/common.go:1069]: Missing the suite name here.

File: tests/e2e/serial/main_test.go [hunk 22]
```diff
@@ -1408,6 +1409,183 @@ func TestKubeletConfigRemediation(t *testing.T) {
...
+
+	if err := f.WaitForScanSettingBindingStatus(f.OperatorNamespace, bindingName, compv1alpha1.ScanSettingBindingPhaseReady); err != nil {
+		t.Fatalf("ScanSettingBinding %s failed to suspend", bindingName)
+	}
+	if err := f.AssertCronJobIsNotSuspended(compsuitectrl.GetRerunnerName(suite.Name)); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestSuspendScaSettingDoesNotCreateScan(t *testing.T) {
```
Comment (rhmdnd) [inline tests/e2e/serial/main_test.go:1511]: Need to debug this test a bit more:

```
=== RUN   TestSuspendScaSettingDoesNotCreateScan
2023/09/27 02:39:28 ScanSettingBinding status SUSPENDED
    main_test.go:1563: Failed to assert ComplianceSuite %s does not exist.
--- FAIL: TestSuspendScaSettingDoesNotCreateScan (5.10s)
```

File: pkg/controller/scansettingbinding/scansettingbinding_controller.go (hunk 18 above)
Comment (rhmdnd) [reply pkg/controller/scansettingbinding/scansettingbinding_controller.go:845]: I expanded the conditional here to include checking for the scansetting changes because without it the CronJob isn't updated because the it's not detected (line 314). Checking this again though.

File: pkg/controller/scansettingbinding/scansettingbinding_controller.go (hunk 20 above)
Comment (rhmdnd) [reply pkg/controller/scansettingbinding/scansettingbinding_controller.go]: Good catch - fixed.

**Comment by rhmdnd** [comment]
@yuumasato @Vincent056 should be ready for another round of reviews.

File: pkg/controller/scansettingbinding/scansettingbinding_controller.go [hunk 23]
```diff
@@ -211,6 +214,51 @@ func (r *ReconcileScanSettingBinding) Reconcile(ctx context.Context, request rec
...
+			// ComplianceSuite or kick off any scans, yet. If or
+			// when the ScanSetting is activated, we will proceed
+			// like normal and create the ComplianceSuite and its
+			// ComplianceScans.
+			if errors.IsNotFound(err) {
+				return reconcile.Result{}, nil
+			}
+
+		} else if !scanSetting.ComplianceSuiteSettings.Suspend {
+			ssb.Status.SetConditionReady()
```
Comment (yuumasato) [inline pkg/controller/scansettingbinding/scansettingbinding_controller.go:247]: Analogous to the log that the `ScanSettingBinding` was suspended, I miss a log entry that the SSB  was resumed.
`log.Info("Resuming ScanSettingBinding", "ScanSettingBinding", ssb.Name)`

File: cmd/manager/operator.go [hunk 24]
```diff
@@ -553,6 +553,9 @@ func ensureDefaultScanSettings(
 			},
 			ComplianceSuiteSettings: compv1alpha1.ComplianceSuiteSettings{
 				Schedule: defaultScanSettingsSchedule,
+				// Find a to make this default to `False`
```
[the remark this answers is not in the document]
Comment (rhmdnd) [reply cmd/manager/operator.go:556]: Aha - I meant to say find a "way" to make this default to false, but I actually addressed that already with line 87 in pkg/apis/compliance/v1alpha1/compliancesuite_types.go.

I'll clean this up.

File: cmd/manager/operator.go [hunk 25]
```diff
@@ -579,6 +582,9 @@ func ensureDefaultScanSettings(
 				AutoApplyRemediations:  true,
 				AutoUpdateRemediations: true,
 				Schedule:               defaultScanSettingsSchedule,
+				// Find a to make this default to `False`
+				// instead of needing to set the value here.
+				Suspend: false,
```
Comment (rhmdnd) [inline cmd/manager/operator.go:587]: Remove this since it's done by default.

File: pkg/controller/scansettingbinding/scansettingbinding_controller.go (hunk 23 above)
Comment (rhmdnd) [reply pkg/controller/scansettingbinding/scansettingbinding_controller.go:247]: Good find.

File: pkg/controller/scansettingbinding/scansettingbinding_controller.go (hunk 18 above)
Comment (rhmdnd) [reply pkg/controller/scansettingbinding/scansettingbinding_controller.go:832]: I was able to revert this and the tests passed, so the new scan settings are getting copied here.

File: doc/usage.md [hunk 26]
```diff
@@ -293,6 +293,53 @@ or `ocp4-var-role-worker`:
 `oc get ccr -n openshift-compliance -o yaml | jq '.items[] | select(.valuesUsed | contains("ocp4-var-role-master") or contains("ocp4-var-role-worker"))'`
 
 
+## Suspending and resuming scan schedules
```
Comment (sheriff-rh) [inline doc/usage.md:296]: This is excellent. Thanks @rhmdnd !

File: pkg/controller/scansettingbinding/scansettingbinding_controller.go [hunk 27]
```diff
@@ -211,6 +214,52 @@ func (r *ReconcileScanSettingBinding) Reconcile(ctx context.Context, request rec
...
 	}
 
+	scanSetting := &compliancev1alpha1.ScanSetting{}
+	if instance.SettingsRef != nil {
+		ssb := instance.DeepCopy()
+		err = r.Client.Get(context.TODO(), types.NamespacedName{Name: instance.SettingsRef.Name, Namespace: request.Namespace}, scanSetting)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("Failed to get ScanSetting: %s", err)
+		}
+		if scanSetting.ComplianceSuiteSettings.Suspend {
```
Comment (Vincent056) [inline pkg/controller/scansettingbinding/scansettingbinding_controller.go:224]: it might be a nitpicking here, So I wonder if should we set SSB conditions based on whether the current status of the CR matches the desired suspended status? so we don't set the condition status on every reconcile loop.

Comment (Vincent056) [reply pkg/controller/scansettingbinding/scansettingbinding_controller.go:224]: similar to `scanSettingBindingStatusNeedsUpdate`

Comment (Vincent056) [inline pkg/controller/scansettingbinding/scansettingbinding_controller.go:224]: would it be ok if we check `suite.spec.Suspend == true` instead of fetching the scansetting again, since we had them updated at line 211
```
err := r.applyConstraint(instance, &suite, instance.SettingsRef, log)
```

File: pkg/controller/scansettingbinding/scansettingbinding_controller.go [hunk 28]
```diff
@@ -211,6 +214,52 @@ func (r *ReconcileScanSettingBinding) Reconcile(ctx context.Context, request rec
...
+			ssb.Status.SetConditionReady()
+			ssb.Status.Phase = compliancev1alpha1.ScanSettingBindingPhaseReady
+			err := r.Client.Status().Update(context.TODO(), ssb)
+			if err != nil {
+				return reconcile.Result{}, fmt.Errorf("Failed to update ScanSettingBinding status: %s", err)
+			}
+		}
+	}
+
+	instance = &compliancev1alpha1.ScanSettingBinding{}
```
Comment (Vincent056) [inline pkg/controller/scansettingbinding/scansettingbinding_controller.go:257]: so maybe here we first check if the status needs to be updated by comparing the phase with `suite.spec.Suspend`, then if they are not matching, we update and then reconcile, if they do match we do nothing, and then in the next step, if the phase is matching what we wanted, we next check the phase, based on the phase, if the phase is in pausing, we check if there is existing suite exit, if they do we will need to run  `if suiteNeedsUpdate(&suite, &found)` and if they don't exist we do not create suit and reconcile.

Comment (Vincent056) [reply pkg/controller/scansettingbinding/scansettingbinding_controller.go:257]: so if we do that i think we don't have to fetch the ssb instance one more time here

File: pkg/controller/scansettingbinding/scansettingbinding_controller.go (hunk 27 above)
Comment (rhmdnd) [reply pkg/controller/scansettingbinding/scansettingbinding_controller.go:224]: Fixed - good call.

File: pkg/controller/scansettingbinding/scansettingbinding_controller.go (hunk 28 above)
Comment (rhmdnd) [reply pkg/controller/scansettingbinding/scansettingbinding_controller.go]: Nice catch - fixed.

File: pkg/apis/compliance/v1alpha1/scansettingbinding_types.go [hunk 29]
```diff
@@ -76,6 +87,15 @@ func (s *ScanSettingBindingStatus) SetConditionReady() {
 	})
 }
 
+func (s *ScanSettingBindingStatus) SetConditionSuspended() {
+	s.Conditions.SetCondition(Condition{
+		Type:    "Suspended",
+		Status:  corev1.ConditionFalse,
+		Reason:  "Suspended",
+		Message: "The scan setting binding uses a scan setting that is suspended",
+	})
```
Comment (rhmdnd) [inline pkg/apis/compliance/v1alpha1/scansettingbinding_types.go:96]: I suppose this could include removing the `Ready` condition.

Comment (rhmdnd) [reply pkg/apis/compliance/v1alpha1/scansettingbinding_types.go:96]: Or we can use the Ready condition, set it to false, and then put `suspended` for the reason, which looks like it's more consistent with the rest of the controller.

File: pkg/controller/scansettingbinding/scansettingbinding_controller.go [hunk 30]
```diff
@@ -762,5 +802,40 @@ func suiteNeedsUpdate(have, found *compliancev1alpha1.ComplianceSuite) bool {
...
-	return ssb.Status.Conditions.GetCondition("Ready") == nil || ssb.Status.OutputRef == nil || ssb.Status.OutputRef.Name == ""
+	if ssb.Status.Conditions.GetCondition("Ready") == nil {
+		return true
+	} else if ssb.Status.OutputRef == nil {
+		return true
+	} else if ssb.Status.OutputRef.Name == "" {
+		return true
+	} else {
+		return false
+	}
```
Comment (rhmdnd) [inline pkg/controller/scansettingbinding/scansettingbinding_controller.go:813]: This can be reverted since we're not expanding the conditional.

**Comment by xiaojiey** [comment]
Retest with 4.14.0-0.nightly-2023-10-06-234925 + code in the PR. Verification pass.
```
##########Test scenario 1# Check the suspend value in default and default-auto-apply ss:
$ oc get ss default -o=jsonpath={.suspend}
false
$ oc get ss default-auto-apply -o=jsonpath={.suspend}
false
##########Test scenario 2# Create a ss with suspend set to false, then suspend the ssb on demand:
$ cat ss.yaml 
...
schedule: "*/3 * * * *"
strictNodeScan: true
timeout: 30m
$ oc apply -f ss.yaml 
scansetting.compliance.openshift.io/test-suspend created
$ oc compliance bind -N test-suspend -S test-suspend profile/ocp4-cis profile/ocp4-cis-node
Creating ScanSettingBinding test-suspend
$ oc get suite -w
NAME           PHASE       RESULT
test-suspend   LAUNCHING   NOT-AVAILABLE
test-suspend   LAUNCHING   NOT-AVAILABLE
test-suspend   LAUNCHING   NOT-AVAILABLE
test-suspend   RUNNING     NOT-AVAILABLE
test-suspend   RUNNING     NOT-AVAILABLE
test-suspend   RUNNING     NOT-AVAILABLE
test-suspend   AGGREGATING   NOT-AVAILABLE
^C
$oc patch ss test-suspend -p '{"suspend":true}' --type='merge'e'
scansetting.compliance.openshift.io/test-suspend patched
$ oc get ssb
NAME           STATUS
test-suspend   SUSPENDED
$ oc get suite -w
NAME           PHASE         RESULT
test-suspend   AGGREGATING   NOT-AVAILABLE
test-suspend   DONE          NON-COMPLIANT
test-suspend   DONE          NON-COMPLIANT
$ oc get cronjob
NAME                    SCHEDULE      SUSPEND   ACTIVE   LAST SCHEDULE   AGE
test-suspend-rerunner   */3 * * * *   True      0        <none>          7m2s

$ oc get job
No resources found in openshift-compliance namespace.
##########Test scenario 3# Create a ss with suspend set to true, then unsuspend the ssb on demand:
cat ss2.yaml
...
schedule: "*/3 * * * *"
showNotApplicable: false
strictNodeScan: true
timeout: 30m
suspend: true
$ oc delete ssb --all
scansettingbinding.compliance.openshift.io "test-suspend" deleted
$ oc apply -f ss2.yaml 
scansetting.compliance.openshift.io/test-suspend-true created
$ oc get ssb test-suspend-true -o=jsonpath={.status} | jq -r
{
  "conditions": [
    {
      "lastTransitionTime": "2023-10-08T11:14:02Z",
      "message": "The scan setting binding uses a scan setting that is suspended",
      "reason": "Suspended",
      "status": "False",
      "type": "Ready"
    }
  ],
  "phase": "SUSPENDED"
}
$ oc patch ss test-suspend-true --patch '{"suspend":false}' --type='merge'
scansetting.compliance.openshift.io/test-suspend-true patched
$ oc get ssb
NAME                STATUS
test-suspend-true   READY
$ oc get suite -w
NAME                PHASE     RESULT
test-suspend-true   RUNNING   NOT-AVAILABLE
...
test-suspend-true   AGGREGATING   NOT-AVAILABLE
test-suspend-true   DONE          NON-COMPLIANT
```

**Review by Vincent056 — APPROVED** [review]
thanks for addressing all the comments,
/lgtm
